rootsanalyzer/rootsanalyzer.go

```diff
@@ -0,0 +1,152 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package rootsanalyzer reports on changes in the set of root certificates
+// returned by a CT Log's get-roots endpoint.
+package rootsanalyzer
+
+import (
+	"bytes"
+	"context"
+	"net/url"
+	"path"
+	"sort"
+	"strings"
+	"text/template"
+
+	"github.com/golang/glog"
+	ct "github.com/google/certificate-transparency-go"
+	"github.com/google/certificate-transparency-go/x509"
+	"github.com/google/monologue/ctlog"
+	"github.com/google/monologue/incident"
+	"github.com/google/monologue/storage"
+)
+
+var (
+	incidentTemplate = template.Must(template.New("roots_incident").Parse(`The root certificates accepted by {{ .Log.Name }} ({{ .Log.URL }}) have changed.
+{{ if gt (len .AddedCerts) 0 }}
+Certificates added ({{ len .AddedCerts }}):
+{{ range .AddedCerts }}{{ .Subject }}
+{{ end }}{{ end }}{{ if gt (len .RemovedCerts) 0 }}
+Certificates removed ({{ len .RemovedCerts }}):
+{{ range .RemovedCerts }}{{ .Subject }}
+{{ end }}{{ end }}`))
+)
+
+type incidentTemplateArgs struct {
+	Log          *ctlog.Log
+	AddedCerts   []*x509.Certificate
+	RemovedCerts []*x509.Certificate
+}
+
+// Run starts an Roots Analyzer, which watches for changes to a CT log's root certificates and creates incident reports for them.
+func Run(ctx context.Context, st storage.RootsReader, rep incident.Reporter, l *ctlog.Log) {
+	rootsChanged, err := st.WatchRoots(ctx, l)
+	if err != nil {
+		glog.Errorf("rootsanalyzer: storage.RootsReader.WatchRoots() = %q", err)
+		return
+	}
+
+	var lastRootSetID storage.RootSetID
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case rootSetID := <-rootsChanged:
+			if lastRootSetID != "" && lastRootSetID != rootSetID {
+				// TODO(RJPercival): If the root set is flapping back to what it recently was, suppress sending an incident report
+				// since it could just be the result of skew between log frontends. However, if it doesn't flap back to the new
+				// root set again within a certain amount of time, then the suppressed report should be sent.
+				oldRoots, err := st.ReadRoots(ctx, lastRootSetID)
+				if err != nil {
+					glog.Errorf("rootsanalyzer: %s", err)
+					return
+				}
+				newRoots, err := st.ReadRoots(ctx, rootSetID)
+				if err != nil {
+					glog.Errorf("rootsanalyzer: %s", err)
+					return
+				}
+				addedCerts, removedCerts := diffRootSets(oldRoots, newRoots)
+				if err := reportChange(ctx, rep, l, addedCerts, removedCerts); err != nil {
+					glog.Errorf("rootsanalyzer: %s", err)
+					return
+				}
+			}
+			lastRootSetID = rootSetID
+		}
+	}
+}
+
+// diffRootSets returns the certificates that have been added or removed in new, relative to old.
+// Neither old nor new are allowed to contain duplicates.
```

Comment:
It'd complicate the code, and it didn't seem necessary since neither root set should contain duplicates. I've added a comment to the RootsWriter interface to mention that it shouldn't store duplicate root certs. I'll add a TODO to handle it though.